crates/fiber-lib/src/rpc/mod.rs

```diff
@@ -89,37 +101,118 @@ pub mod server {
     {
     }
 
-    async fn build_server(addr: &str) -> Server {
-        #[cfg(debug_assertions)]
-        {
-            // Use socket2 to set reuse address and reuse port,
-            // so that we can restart the server without waiting for the port to be released.
-            // it will avoid the error: "Address already in use" in CI.
-            use socket2::{Domain, Socket, Type};
-            let addr = addr.parse().expect("valid address");
-            let domain = Domain::for_address(addr);
-            let socket = Socket::new(domain, Type::STREAM, None).expect("new socket");
-            socket
-                .set_nonblocking(true)
-                .expect("set socket nonblocking");
-            socket.set_reuse_address(true).expect("set reuse address");
-            #[cfg(all(unix, not(any(target_os = "solaris", target_os = "illumos"))))]
-            socket.set_reuse_port(true).expect("set reuse port");
-
-            socket.bind(&addr.into()).expect("bind socket to address");
-            socket.listen(4096).expect("listen socket at the port");
-
-            jsonrpsee::server::Server::builder()
-                .build_from_tcp(socket)
-                .expect("JsonRPC server built from TCP")
-        }
-        #[cfg(not(debug_assertions))]
-        {
-            Server::builder()
-                .build(addr)
-                .await
-                .expect("JsonRPC server built")
+    async fn start_server(
+        addr: &str,
+        auth: Option<BiscuitAuth>,
+        methods: impl Into<Methods>,
+    ) -> Result<(ServerHandle, SocketAddr)> {
+        let listener = TcpListener::bind(addr).await?;
+        let listen_addr = listener.local_addr().expect("get local address");
+
+        // From this example
+        // https://github.com/paritytech/jsonrpsee/blob/d3d9fa8553756751ad913830e7d0d0faca614cb5/examples/examples/jsonrpsee_as_service.rs
+
+        // This state is cloned for every connection
+        // all these types based on Arcs and it should
+        // be relatively cheap to clone them.
+        //
+        // Make sure that nothing expensive is cloned here
+        // when doing this or use an `Arc`.
+        #[derive(Clone)]
+        struct PerConnection<RpcMiddlewave, HttpMiddlewave> {
+            methods: Methods,
+            stop_handle: StopHandle,
+            svc_builder: TowerServiceBuilder<RpcMiddlewave, HttpMiddlewave>,
         }
+
+        // Each RPC call/connection get its own `stop_handle`
+        // to able to determine whether the server has been stopped or not.
+        //
+        // To keep the server running the `server_handle`
+        // must be kept and it can also be used to stop the server.
```

Comment:
Did you confirmed these two things:
1. there is no performance lose if auth is not enabled
2. how much on the performence lose if auth is enabled.